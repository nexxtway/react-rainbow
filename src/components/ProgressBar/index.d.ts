import { BaseProps } from '../types';

export interface ProgressBarProps extends BaseProps {
    value?: number;
    size?: 'x-small' | 'small' | 'medium' | 'large';
    variant?: 'brand' | 'success';
    assistiveText?: string;
}

export default function(props: ProgressBarProps): JSX.Element | null;
