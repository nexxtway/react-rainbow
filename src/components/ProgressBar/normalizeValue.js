export default function normalizeValue(value) {
    if (value < 0) {
        return 0;
    }
    if (value > 100) {
        return 100;
    }
    return value;
}
