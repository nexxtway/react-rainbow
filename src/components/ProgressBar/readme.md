##### ProgressBar base in action

```js
import React from 'react';
import { ProgressBar } from 'react-rainbow-components';

class ProgressBarInAction extends React.Component {
    constructor(props) {
        super(props);
        this.state = { value: 0 };
        this.increment = this.increment.bind(this);
    }

    componentDidMount() {
        this.increment();
    }

    increment() {
        const { value } = this.state;
        if (value === 100) {
            this.setState({ value: 0 });
        } else {
            this.setState({ value: value + 0.5 });
        }
        setTimeout(this.increment, 50);
    }

    render() {
        const { value } = this.state;
        return (
            <div className="rainbow-p-around_x-large">
                <ProgressBar value={value} />
            </div>
        );
    }
}

<ProgressBarInAction />;
```

##### ProgressBar success

```js
import React from 'react';
import { ProgressBar } from 'react-rainbow-components';

<div className="rainbow-p-around_x-large">
    <ProgressBar value={50} variant="success" />
</div>
```

##### descriptive ProgressBar

```js
import React from 'react';
import { ProgressBar } from 'react-rainbow-components';

<div className="rainbow-p-around_x-large">
    <div className="rainbow-align-content_space-between rainbow-p-bottom_x-small">
        <span className="rainbow-font-size-text_medium rainbow-color_gray-4">
            Descriptive ProgressBar
        </span>
        <span aria-hidden="true">
            <strong className="rainbow-font-size-text_medium rainbow-color_brand">
                25% Complete
            </strong>
        </span>
    </div>
    <ProgressBar value={25} />
</div>
```

##### ProgressBar with size variants

```js
import React from 'react';
import { ProgressBar } from 'react-rainbow-components';

<div className="rainbow-p-around_large">
    <div className="rainbow-p-around_medium">
        <ProgressBar value={25} size="x-small" />
        <span className="rainbow-flex rainbow-m-top_x-small rainbow-color_gray-4">
            size: x-small
        </span>
    </div>
    <div className="rainbow-p-around_medium">
        <ProgressBar value={50} size="small" />
        <span className="rainbow-flex rainbow-m-top_x-small rainbow-color_gray-4">size: small</span>
    </div>
    <div className="rainbow-p-around_medium">
        <ProgressBar value={35} size="medium" />
        <span className="rainbow-flex rainbow-m-top_x-small rainbow-color_gray-4">
            size: medium
        </span>
    </div>
    <div className="rainbow-p-around_medium">
        <ProgressBar value={75} size="large" />
        <span className="rainbow-flex rainbow-m-top_x-small rainbow-color_gray-4">size: large</span>
    </div>
</div>
```
