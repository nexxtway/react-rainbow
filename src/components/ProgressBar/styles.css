.rainbow-progress-bar {
  display: block;
  background: rgba(0, 184, 245, 0.2);
  height: 0.5rem;
  width: 100%;
  border: 0;
  position: relative;
  border-radius: 0.5rem; }

.rainbow-progress-bar--success {
  background: rgba(29, 233, 182, 0.2); }

.rainbow-progress-bar--x-small {
  height: 0.125rem; }

.rainbow-progress-bar--small {
  height: 0.25rem; }

.rainbow-progress-bar--medium {
  height: 0.625rem; }

.rainbow-progress-bar--large {
  height: 1rem; }

.rainbow-progress-bar_value {
  display: block;
  background: #01b6f5;
  height: 100%; }

.rainbow-progress-bar_value--success {
  background: #1de9b6; }

.rainbow-progress-bar .rainbow-progress-bar_value {
  border-radius: 1rem; }
