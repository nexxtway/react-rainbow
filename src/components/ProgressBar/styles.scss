@import "../../styles/colors";

.rainbow-progress-bar {
    display: block;
    width: 100%;
    height: 0.5rem;
    background: $color-brand-light;
    border: 0;
    position: relative;
}

.rainbow-progress-bar_success {
    background: $color-success-light;
}

.rainbow-progress-bar_x-small {
    height: 0.125rem;
}

.rainbow-progress-bar_small {
    height: 0.25rem;
}

.rainbow-progress-bar_medium {
    height: 0.625rem;
}

.rainbow-progress-bar_large {
    height: 1rem;
}

.rainbow-progress-bar_circular {
    border-radius: 0.5rem;
}

.rainbow-progress-bar__value {
    display: block;
    background: $color-brand;
    height: 100%;
}

.rainbow-progress-bar__value_success {
    background: $color-success;
}

.rainbow-progress-bar_circular .rainbow-progress-bar__value {
    border-radius: 0.5rem;
}

.rainbow-progress-bar_vertical {
    height: 100%;
    width: 0.5rem;
}

.rainbow-progress-bar_vertical.rainbow-progress-bar_x-small {
    width: 0.125rem;
}

.rainbow-progress-bar_vertical.rainbow-progress-bar_small {
    width: 0.25rem;
}

.rainbow-progress-bar_vertical.rainbow-progress-bar_medium {
    width: 0.625rem;
}

.rainbow-progress-bar_vertical.rainbow-progress-bar_large {
    width: 1rem;
}
