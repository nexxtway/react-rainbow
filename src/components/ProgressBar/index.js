import React from 'react';
import PropTypes from 'prop-types';
import AsistiveText from './../AssistiveText';
import normalizeValue from './normalizeValue';
import StyledContainer from './styled/container';
import StyledBar from './styled/bar';

/**
 * Progress bar component communicates to the user the progress of a particular process.
 */
export default function ProgressBar(props) {
    const { className, style, assistiveText, value, size, variant } = props;

    const normalizedValue = normalizeValue(value);
    const WIDTH = { width: `${normalizedValue}%` };

    return (
        <StyledContainer
            className={className}
            style={style}
            size={size}
            variant={variant}
            aria-valuemin="0"
            aria-valuemax="100"
            aria-valuenow={normalizedValue}
            role="progressbar"
        >
            <StyledBar variant={variant} style={WIDTH}>
                <AsistiveText text={assistiveText} />
            </StyledBar>
        </StyledContainer>
    );
}

ProgressBar.propTypes = {
    /** The percentage value of the progress bar. It must be between 0 and 100. */
    value: PropTypes.number,
    /** The size of the progress bar. Valid values are x-small, small, medium, and large.
     * The default value is medium. */
    size: PropTypes.oneOf(['x-small', 'small', 'medium', 'large']),
    /** The variant of the progress bar. Valid values are brand and success. */
    variant: PropTypes.oneOf(['brand', 'success']),
    /** A description for assistive sreen readers. */
    assistiveText: PropTypes.string,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** An object with custom style applied to the outer element. */
    style: PropTypes.object,
};

ProgressBar.defaultProps = {
    value: 0,
    size: 'medium',
    variant: 'brand',
    assistiveText: '',
    className: undefined,
    style: undefined,
};
