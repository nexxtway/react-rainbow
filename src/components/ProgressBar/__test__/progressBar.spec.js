import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ProgressBar from './../index';

describe('<ProgressBar/>', () => {
    it('should has set the value passed', () => {
        const component = shallow(<ProgressBar value={25} />);

        expect(component.find('span.rainbow-progress-bar__value').prop('style').width).toBe('25%');
    });
    it('should be accesible', () => {
        const component = shallow(<ProgressBar value={25} />);

        expect(component.find('div[role="progressbar"]').exists()).toBe(true);
        expect(component.find('div[aria-valuemax="100"]').exists()).toBe(true);
        expect(component.find('div[aria-valuemin="0"]').exists()).toBe(true);
        expect(component.find('div[aria-valuenow=25]').exists()).toBe(true);
    });
    it('should have the right class names when variant, color and size default and have a custom class', () => {
        const component = renderer.create(
            <ProgressBar className="custom-class" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant circular', () => {
        const component = renderer.create(
            <ProgressBar variant="circular" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when size large', () => {
        const component = renderer.create(
            <ProgressBar size="large" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when size medium', () => {
        const component = renderer.create(
            <ProgressBar size="medium" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when size small', () => {
        const component = renderer.create(
            <ProgressBar size="small" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when size x-small', () => {
        const component = renderer.create(
            <ProgressBar size="x-small" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when color success', () => {
        const component = renderer.create(
            <ProgressBar color="success" />,
        );
        expect(component).toMatchSnapshot();
    });
});
