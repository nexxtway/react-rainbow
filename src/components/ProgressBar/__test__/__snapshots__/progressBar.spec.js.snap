// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar/> should have the right class names when color success 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_medium rainbow-progress-bar_success"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value rainbow-progress-bar__value_success"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when size large 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_large"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when size medium 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_medium"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when size small 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_small"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when size x-small 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_x-small"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when variant circular 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_medium rainbow-progress-bar_circular"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;

exports[`<ProgressBar/> should have the right class names when variant, color and size default and have a custom class 1`] = `
<div
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow={0}
  className="rainbow-progress-bar rainbow-progress-bar_medium custom-class"
  role="progressbar"
>
  <span
    className="rainbow-progress-bar__value"
    style={
      Object {
        "width": "0%",
      }
    }
  />
</div>
`;
